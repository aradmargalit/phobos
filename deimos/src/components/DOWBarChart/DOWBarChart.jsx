/* eslint-disable react/no-array-index-key */
import { sortBy as _sortBy } from 'lodash';
import React from 'react';
import { Bar, BarChart, Cell, LabelList, XAxis, YAxis } from 'recharts';

import AngledGraphTick from '../AngledGraphTick';

const COLORS = ['#f7bdbc', '#f49d9a', '#f07c79', '#ec5b57', '#d4524e', '#bd4946', '#8e3734'];

const abbreviate = (dayBreakdown) =>
  dayBreakdown.map((day) => ({ ...day, day_of_week: day.day_of_week.slice(0, 3) }));

export default function DOWBarChart({ dayBreakdown }) {
  const sortedDays = _sortBy(dayBreakdown, 'count');

  // Find the proportion of each day, and give it a darker color if it's higher
  const calculateColor = (day) => COLORS[sortedDays.indexOf(day)];

  return (
    <div className="statistics--dow">
      <h3>Daily Breakdown</h3>
      <BarChart data={abbreviate(dayBreakdown)} width={500} height={300} margin={{ top: 20 }}>
        <XAxis dataKey="day_of_week" height={80} tick={<AngledGraphTick />} />
        <YAxis />
        <Bar dataKey="count" fill="#0e5a6d">
          <LabelList dataKey="count" position="top" />
          {/* This is exclusively for alternating colors in the graph */}
          {dayBreakdown.map((day) => (
            <Cell key={day.day_of_week} fill={calculateColor(day)} />
          ))}
        </Bar>
      </BarChart>
    </div>
  );
}
