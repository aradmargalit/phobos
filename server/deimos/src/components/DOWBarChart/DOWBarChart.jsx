/* eslint-disable react/no-array-index-key */
import React from 'react';
import {
  Bar,
  BarChart,
  Cell,
  LabelList,
  ResponsiveContainer,
  XAxis,
  YAxis,
} from 'recharts';

import AngledGraphTick from '../AngledGraphTick';

export default function DOWBarChart({ colors, dayBreakdown }) {
  return (
    <div className="statistics--dow">
      <h3>Daily Breakdown</h3>
      <ResponsiveContainer width={500} height={300}>
        <BarChart
          data={dayBreakdown}
          margin={{
            top: 5,
            right: 30,
            left: 20,
            bottom: 5,
          }}
        >
          <XAxis
            dataKey="day_of_week"
            height={120}
            tick={<AngledGraphTick />}
          />
          <YAxis />
          <Bar dataKey="count" fill="#0e5a6d">
            <LabelList dataKey="count" position="top" />
            {dayBreakdown.map((entry, index) => (
              <Cell
                key={`cell-${index}`}
                fill={colors[index % colors.length]}
              />
            ))}
          </Bar>
        </BarChart>
      </ResponsiveContainer>
    </div>
  );
}
