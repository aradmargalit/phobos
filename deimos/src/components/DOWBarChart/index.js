import DOWBarChart from './DOWBarChart';

export default DOWBarChart;
