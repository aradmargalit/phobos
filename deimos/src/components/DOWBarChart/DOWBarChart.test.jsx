import { shallow } from 'enzyme';
import React from 'react';

import DOWBarChart from './DOWBarChart';

const generateComponent = (props) => {
  const defaultProps = {
    dayBreakdown: [{ day_of_week: 'Monday', count: 10 }],
  };
  const mergedProps = { ...defaultProps, ...props };

  return shallow(<DOWBarChart {...mergedProps} />);
};

describe('<DOWBarChart />', () => {
  it('renders correctly', () => {
    expect(generateComponent()).toMatchSnapshot();
  });
});
