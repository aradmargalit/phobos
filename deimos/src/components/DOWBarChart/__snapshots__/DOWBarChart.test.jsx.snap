// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DOWBarChart /> renders correctly 1`] = `
<div
  className="statistics--dow"
>
  <h3>
    Daily Breakdown
  </h3>
  <ResponsiveContainer
    debounce={0}
    height={400}
    width="100%"
  >
    <BarChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "count": 10,
            "day_of_week": "Monday",
          },
        ]
      }
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 20,
          "right": 20,
          "top": 20,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
    >
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="day_of_week"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={120}
        hide={false}
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tick={<AngledGraphTick />}
        tickCount={5}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={0}
        hide={false}
        mirror={false}
        orientation="left"
        padding={
          Object {
            "bottom": 0,
            "top": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        type="number"
        width={60}
        yAxisId={0}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        data={Array []}
        dataKey="count"
        fill="#0e5a6d"
        hide={false}
        isAnimationActive={true}
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        xAxisId={0}
        yAxisId={0}
      >
        <LabelList
          dataKey="count"
          position="top"
          valueAccessor={[Function]}
        />
        <Cell
          fill="#f7bdbc"
          key="Monday"
        />
      </Bar>
    </BarChart>
  </ResponsiveContainer>
</div>
`;
